### PR TITLE
Adapt logic for global.conf changes

### PR DESCRIPTION
**Pull request by jschmid1**
Signed-off-by: Joshua Schmid <jschmid@suse.de>

* changed logic for dependency resolution
* global.conf changes now result in core service restart(flags) (mon, mgr, osd)

~Ignore whitespace and indentation fixes. [auto]~

suse_internal: bsc#1149400

-----------------

**Checklist:**
~- [ ] Added unittests and or functional tests~
~- [ ] Adapted documentation~
- [x] Referenced issues or internal bugtracker
- [ ] Ran integration tests successfully (trigger with "@susebot run teuthology" in a GitHub comment; see the [wiki](https://github.com/SUSE/DeepSea/wiki/Testing) for more information)

### CODE REVIEW

File: srv/modules/runners/changed.py [hunk 1]
```diff
@@ -74,6 +78,31 @@ def dependencies(self):
...
+        self._depends = role
+
+    def _set_depends(self):
+        """ Set dependency to self if not global """
+        if self._role_name == 'global':
+            # if the global config changed
+            # stage restart on the core services.
+            self.add_dependencies(Role(role_name='mon'))
+            self.add_dependencies(Role(role_name='mgr'))
+            self.add_dependencies(Role(role_name='storage'))
```
Comment (jan--f) [inline srv/modules/runners/changed.py]: `mds` too?

Comment (jschmid1) [reply srv/modules/runners/changed.py]: I limited the restarts to the core services, since the mds should have it's own section (mds.conf)

Comment (swiftgist) [reply srv/modules/runners/changed.py]: While that's true, I don't know if any of us know of all the Ceph variables that could impact both mds and another component.  If an admin updates the global.conf not quite understanding that it's more suited for mds.conf, I would still expect DeepSea to restart everything.  Currently, the mon, mgr and osds will restart, but not the mds.  So, the admin will still have suffer the penalty of the serial restart, but still have to investigate what went wrong.

File: srv/modules/runners/changed.py [hunk 2]
```diff
@@ -102,9 +131,11 @@ def rgw_configurations(self):
...
         roles = []
         try:
-            roles = list(local.cmd("I@roles:master", 'pillar.get',
-                                   ['rgw_configurations'],
-                                   tgt_type="compound").values())[0]
+            roles = list(
+                local.cmd(
+                    "I@roles:master",
+                    'pillar.get', ['rgw_configurations'],
+                    tgt_type="compound").values())[0]
```
Comment (mgfritch) [inline srv/modules/runners/changed.py]: Why the change in whitespace?

File: srv/modules/runners/changed.py [hunk 3]
```diff
@@ -129,8 +160,8 @@ def __init__(self, **kwargs):
         self.base_dir = '/srv/salt/ceph/configuration/files/'
         self.checksum_dir = self.base_dir + 'ceph.conf.checksum/'
         self.checksum_file = self.checksum_dir + self.role.conf_filename + self.role.conf_extension
-        log.debug("dependencies of role {}: {}".format(self.role.name,
-                                                       self.role.dependencies_unwrapped()))
+        log.debug("dependencies of role {}: {}".format(
+            self.role.name, self.role.dependencies_unwrapped()))
```
Comment (mgfritch) [inline srv/modules/runners/changed.py]: Change in whitespace to stay under char limit per line?

File: srv/modules/runners/changed.py (hunk 1 above)
Comment (jschmid1) [reply srv/modules/runners/changed.py]: I get the point, but do we also stage 'rgws' for restarts then?

Comment (swiftgist) [reply srv/modules/runners/changed.py]: I think so, which then gets us into `rgw_configurations`.  sigh... Considering that's not a one-liner, should that just be a separate PR.  I don't know the urgency of this patch currently.

Comment (jan--f) [reply srv/modules/runners/changed.py]: Yeah I think the issue is that _any_ config option can potentially be in the global section. Techincally we'd have to restart the whole cluster on a change here.

This is certainly very heavy-handed. Not sure what the reasonable way is.

Comment (swiftgist) [reply srv/modules/runners/changed.py]: It is, but I swear that's where we were with this originally.  Since Ceph literally has 1K+ parameters (and maybe users shouldn't use them, but that doesn't stop them), we have no map to say `global X` impacts components A and B.  And even if we did, our maintained version would likely be wrong the minute we shipped.

Restarting the whole cluster gracefully is expensive, but anything less results in a bug like this one.

**Comment by jschmid1** [comment]
note to self: split up commits in

* whitespace changes
* actual code changes

File: srv/modules/runners/changed.py (hunk 1 above)
Comment (jschmid1) [reply srv/modules/runners/changed.py]: Changed it to also restart `rgw` and `mds` - `igw` is not maintained in that runner. However I'd argue this is the right interface to stage restarts.

File: srv/modules/runners/changed.py (hunk 2 above)
Comment (jschmid1) [reply srv/modules/runners/changed.py]: I reverted the style changes

File: srv/modules/runners/changed.py (hunk 3 above)
Comment (jschmid1) [reply srv/modules/runners/changed.py]: I reverted the style changes

**Comment by jschmid1** [comment]
> note to self: split up commits in
> 
~> * whitespace changes~
> * actual code changes

For now, let's stick to the actual code changes until we reached consensus on the style-guide

**Comment by jschmid1** [comment]
@susebot run teuthology

**Comment by jschmid1** [comment]
> Commit [edbb5ad](https://github.com/SUSE/DeepSea/commit/edbb5ad2381332027aac7b9268bbd8c768fbd773) is _NOT OK_ for suite deepsea:tier2.
> Check tests results in the Jenkins job: http://ci.ses.suse.de:8080/job/pr-deepsea/282/

the CI caught a bug in this patch, fixing

File: srv/modules/runners/changed.py [hunk 4]
```diff
@@ -74,6 +75,34 @@ def dependencies(self):
...
+        """ Set dependency to self if not global """
+        if self._role_name == 'global':
+            # if the global config changed
+            # stage restart on the core services
+            self.add_dependencies(Role(role_name='mon'))
+            self.add_dependencies(Role(role_name='mgr'))
+            self.add_dependencies(Role(role_name='storage'))
+            # also stage restarts on gateways (igw is not maintained here)
+            self.add_dependencies(Role(role_name='mds'))
+            self.add_dependencies(Role(role_name='rgw'))
```
Comment (jschmid1) [inline srv/modules/runners/changed.py]: This causes a loop due to this https://github.com/SUSE/DeepSea/blob/edbb5ad2381332027aac7b9268bbd8c768fbd773/srv/modules/runners/changed.py#L125-L145

I'm not sure why we set the dependency for `global` on a rgw_configuration change?. I'd argue for getting rid of this

Comment (jschmid1) [reply srv/modules/runners/changed.py]: by the looks of it, we don't need `rgw_configurations` at all. According to 

https://github.com/SUSE/DeepSea/blob/master/srv/salt/ceph/rgw/restart/controlled/default.sls#L1-L10

we restart all `rgw_configurations` if a `restart_rgw` grain is set. No need to check rgw_configrations inside the `changed` runner at all.

**Comment by jschmid1** [comment]
ready for a second round of reviews

**Comment by jschmid1** [comment]
@susebot run teuthology

**Comment by jschmid1** [comment]
unrelated to *this* pr, but it seems we have a failure in stage.4.. re-running to verify

**Comment by jschmid1** [comment]
@susebot run teuthology

**Comment by jschmid1** [comment]
@smithfarm It seems that the CI is currently not able to deploy nfs-ganesha. Did we update the build artifacts already?

**Comment by jschmid1** [comment]
@susebot run teuthology

**Comment by smithfarm** [comment]
@jschmid1 I don't see where the CI is unable to deploy nfs-ganesha. What I see is this:

* Stage 4 completes successfully
* ganesha smoke test fails with:

```
2019-09-16T14:00:38.503 INFO:teuthology.orchestra.run.target192168000079.stderr:scripts/ganesha_smoke_test.sh: line 225: /root/mnt/rgw/bucket1/hello.txt: No such file or directory
```

**Comment by jschmid1** [comment]
The failure is weird though. All tests before and after were successful.. This branch is clearly not changing anything in the NFS-ganesha area.. re-triggering to be extra-extra-sure

**Comment by jschmid1** [comment]
@susebot run teuthology

**Comment by jschmid1** [comment]
apparently we have two nfs-ganesha tests:

1) with mds backend
2) with rgw backed

only 1) fails. Maybe the mds state fails while checking for changed files somehow?
Why wouldn't the standard(non-ganesha) mds test fail?..

**Comment by jschmid1** [comment]
I'm puzzled..

**Comment by jschmid1** [comment]
Closing.. Reopen if needed
